sycl/unittests/helpers/PiMock.hpp

```diff
@@ -0,0 +1,197 @@
+/// redefinitions would also affect other platforms' behavior.
+/// Therefore, any plugin-related information is fully copied whenever
+/// a user-passed SYCL object instance is being mocked.
+/// The underlying SYCL platform must be a non-host plaftorm to facilitate
+/// plugin usage.
+///
+/// Simple usage examples would look like this:
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// unittest::PiMock Mock;
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// platform &MockP = Mock.getPlatform();
+/// /*...*/
+/// ```
+/// ```
+/// pi_result redefinePiProgramRetain(pi_program program) { /*code*/ }
+/// /*...*/
+/// queue Q;
+/// unittest::PiMock Mock(Q);
+/// Mock.redefine<PiApiKind::piProgramRetain>(redefinePiProgramRetain);
+/// Q.submit(/* expecting mock behavior */);
+/// /*...*/
+/// ```
+class PiMock {
+public:
+  /// Default-constructs PiMock.
+  ///
+  /// Delegates to the device_selector-based constructor by passing
+  /// a default_constructor and assuming a non-host device will be
+  /// available.
+  PiMock() : PiMock(cl::sycl::default_selector{}) {}
+
+  /// Constructs PiMock from a queue.
+  ///
+  /// \param Queue is a reference to a SYCL queue to which
+  ///        the mock redefinitions will apply.
+  explicit PiMock(cl::sycl::queue &Queue)
+      : PiMock(Queue.get_device().get_platform()) {}
+
+  /// Constructs PiMock from a reference to a SYCL platform instance.
+  ///
+  /// A new plugin will be stored into the platform instance, which
+  /// will no longer share the plugin with other platform instances
+  /// within the given context. A separate platform instance will be
+  /// held by the PiMock instance.
+  ///
+  /// \param OriginalPlatform is a reference to a SYCL platform.
+  explicit PiMock(const cl::sycl::platform &OriginalPlatform) {
+    assert(!OriginalPlatform.is_host() && "PI mock isn't supported for host");
+    // Extract impl and plugin handles
+    auto ImplPtr = detail::getSyclObjImpl(OriginalPlatform);
+    const RT::PiPlugin &OriginalPiPlugin = ImplPtr->getPlugin().getPiPlugin();
+    // Copy the PiPlugin, thus untying our to-be mock platform from other
+    // platforms within the context. Reset our platform to use the new plugin.
+    auto NewPluginPtr = std::make_shared<detail::plugin>(OriginalPiPlugin);
+    ImplPtr->setPlugin(NewPluginPtr);
+    // Extract the new PiPlugin instance by a non-const pointer,
+    // explicitly allowing modification
+    PiPluginMockPtr = &const_cast<RT::PiPlugin &>(NewPluginPtr->getPiPlugin());
+    // Save a copy of the platform resource
+    MPlatform = OriginalPlatform;
+  }
+
+  /// Constructs PiMock from a device_selector, provided that
+  /// a non-host device can and will be selected.
+  ///
+  /// \param DevSelector is a reference to a device_selector instance.
+  explicit PiMock(const cl::sycl::device_selector &DevSelector)
```

Comment:
Suggest calling `explicit PiMock(const cl::sycl::platform &OriginalPlatform) {` from this c'tor.

Reply:
Done